netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersDecoder.java

```diff
@@ -0,0 +1,498 @@
+
+    GrpcHttp2InboundHeaders(int numHeadersGuess) {
+      namesAndValues = new byte[numHeadersGuess * 2][];
+      values = new AsciiString[numHeadersGuess];
+    }
+
+    protected Http2Headers add(AsciiString name, AsciiString value) {
+      if (namesAndValuesIdx == namesAndValues.length) {
+        expandHeadersAndValues();
+      }
+      final byte[] nameBytes = bytes(name);
+      final byte[] valueBytes = toRawSeralizedHeader(nameBytes, bytes(value));
+      values[namesAndValuesIdx / 2] = value;
+      namesAndValues[namesAndValuesIdx] = nameBytes;
+      namesAndValuesIdx++;
+      namesAndValues[namesAndValuesIdx] = valueBytes;
+      namesAndValuesIdx++;
+      return this;
+    }
+
+    protected CharSequence get(AsciiString name) {
+      for (int i = 0; i < namesAndValuesIdx; i += 2) {
+        if (equals(name, namesAndValues[i])) {
+          return values[i / 2];
+        }
+      }
+      return null;
+    }
+
+    @Override
+    public List<CharSequence> getAll(CharSequence csName) {
+      final AsciiString name = requireAsciiString(csName);
+      final List<CharSequence> returnValues = new ArrayList<CharSequence>(4);
+      for (int i = 0; i < namesAndValuesIdx; i += 2) {
+        if (equals(name, namesAndValues[i])) {
+          returnValues.add(values[i / 2]);
+        }
+      }
+      return returnValues;
+    }
+
+    /**
+     * Returns the header names and values as bytes. An even numbered index contains the
+     * {@code byte[]} representation of a header name (in insertion order), and the subsequent
+     * odd index number contains the corresponding header value.
+     *
+     * <p>The values of binary headers (with a -bin suffix), are already base64 decoded.
+     *
+     * <p>The array may contain several {@code null} values at the end. A {@code null} value an
+     * index means that all higher numbered indices also contain {@code null} values.
+     */
+    byte[][] namesAndValues() {
+      return namesAndValues;
+    }
+
+    /**
+     * Returns the number of none-null headers in {@link #namesAndValues()}.
+     */
+    protected int numHeaders() {
+      return namesAndValuesIdx / 2;
+    }
+
+    protected static boolean equals(AsciiString str0, byte[] str1) {
+      return equals(str0.array(), str0.arrayOffset(), str0.length(), str1, 0, str1.length);
+    }
+
+    protected static boolean equals(AsciiString str0, AsciiString str1) {
+      return equals(str0.array(), str0.arrayOffset(), str0.length(), str1.array(),
+          str1.arrayOffset(), str1.length());
+    }
+
+    protected static boolean equals(byte[] bytes0, int offset0, int length0, byte[] bytes1,
+        int offset1, int length1) {
+      if (length0 != length1) {
+        return false;
+      }
+      return PlatformDependent.equals(bytes0, offset0, bytes1, offset1, length0);
+    }
+
+    protected static byte[] bytes(AsciiString str) {
+      return str.isEntireArrayUsed() ? str.array() : str.toByteArray();
```

Comment:
You might check if this is ever true.  The changes you made in https://github.com/grpc/grpc-java/commit/edf2c62fd55b74be7b1a85191f8e2f27099fdf75  I think cause huffman coded headers to always over allocate (by roundabout way  of HuffmanDecoder constructor argument).

It may not be a problem, but it would be worth checking that this works as expected.